ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkId.java

```diff
@@ -0,0 +1,378 @@
+      }
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Non EIP-2124 behaviour
+   *
+   * @param peerGenesisOrCheckSumHash Hash or checksum to be validated.
+   * @return boolean
+   */
+  public boolean peerCheck(final Bytes32 peerGenesisOrCheckSumHash) {
+    return !peerGenesisOrCheckSumHash.equals(genesisHash);
+  }
+
+  private boolean isHashKnown(final String forkHash) {
+    for (ForkIdEntry j : forkAndHashList) {
+      if (forkHash.equals(j.hash)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isForkKnown(final Long nextFork) {
+    if (highestKnownFork < nextFork) {
+      return true;
+    }
+    for (ForkIdEntry j : forkAndHashList) {
+      if (nextFork.equals(j.next)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isRemoteAwareOfPresent(final String forkHash, final Long nextFork) {
+    for (ForkIdEntry j : forkAndHashList) {
+      if (forkHash.equals(j.hash)) {
+        if (nextFork.equals(j.next)) {
+          return true;
+        } else if (j.next == 0L) {
+          return highestKnownFork <= nextFork; // Remote aware of future fork
+        } else {
+          return false;
+        }
+      }
+    }
+    return false;
+  }
+
+  private ArrayDeque<ForkIdEntry> collectForksAndHashes(
+      final Set<Long> forks, final Long currentHead) {
+    boolean first = true;
+    ArrayDeque<ForkIdEntry> forkList = new ArrayDeque<>();
+    Iterator<Long> iterator = forks.iterator();
+    while (iterator.hasNext()) {
+      Long forkBlockNumber = iterator.next();
+      if (highestKnownFork < forkBlockNumber) {
+        highestKnownFork = forkBlockNumber;
+      }
+      if (first) {
+        // first fork
+        first = false;
+        forkList.add(
+            new ForkIdEntry(
+                updateCrc(this.genesisHash.getHexString()), forkBlockNumber)); // Genesis
+        updateCrc(forkBlockNumber);
+
+      } else if (!iterator.hasNext()) {
+        // most recent fork
+        forkList.add(new ForkIdEntry(getCurrentCrcHash(), forkBlockNumber));
+        updateCrc(forkBlockNumber);
+        lastKnownEntry = new ForkIdEntry(getCurrentCrcHash(), 0L);
+        forkList.add(lastKnownEntry);
+        if (currentHead > forkBlockNumber) {
+          this.forkNext = 0L;
+        } else {
+          this.forkNext = forkBlockNumber;
+        }
+
+      } else {
+        forkList.add(new ForkIdEntry(getCurrentCrcHash(), forkBlockNumber));
+        updateCrc(forkBlockNumber);
+      }
+    }
+    return forkList;
+  }
+
+  private String updateCrc(final Long block) {
+    byte[] byteRepresentationFork = longToBigEndian(block);
+    crc.update(byteRepresentationFork, 0, byteRepresentationFork.length);
+    return getCurrentCrcHash();
+  }
+
+  private String updateCrc(final String hash) {
+    byte[] byteRepresentation = hexStringToByteArray(hash);
+    crc.update(byteRepresentation, 0, byteRepresentation.length);
+    return getCurrentCrcHash();
+  }
```

Comment:
Would it be possible to keep these as `BytesValue`s, all the way through instead of `String`s?